crates/sui-graphql-rpc/src/server/builder.rs

```diff
@@ -40,49 +43,108 @@ use http::{HeaderValue, Method, Request};
 use hyper::server::conn::AddrIncoming as HyperAddrIncoming;
 use hyper::Body;
 use hyper::Server as HyperServer;
+use mysten_metrics::spawn_monitored_task;
 use mysten_network::callback::{CallbackLayer, MakeCallbackHandler, ResponseHandler};
 use std::convert::Infallible;
 use std::net::TcpStream;
+use std::sync::atomic::{AtomicU64, Ordering::Relaxed};
+use std::sync::Arc;
 use std::{any::Any, net::SocketAddr, time::Instant};
 use sui_graphql_rpc_headers::{LIMITS_HEADER, VERSION_HEADER};
 use sui_package_resolver::{PackageStoreWithLruCache, Resolver};
 use sui_sdk::SuiClientBuilder;
 use tokio::sync::OnceCell;
+use tokio_util::sync::CancellationToken;
 use tower::{Layer, Service};
 use tower_http::cors::{AllowOrigin, CorsLayer};
-use tracing::{info, warn};
+use tracing::{error, info, warn};
 use uuid::Uuid;
 
 pub struct Server {
     pub server: HyperServer<HyperAddrIncoming, IntoMakeServiceWithConnectInfo<Router, SocketAddr>>,
+    /// The following fields are exposed for background tasks
+    pub checkpoint_watermark: Arc<AtomicU64>,
+    pub(crate) state: AppState,
+    pub(crate) db_reader: Db,
 }
 
 impl Server {
+    /// Start the GraphQL service and any background tasks it is dependent on. When a cancellation
+    /// signal is received, the method waits for all tasks to complete before returning.
     pub async fn run(self) -> Result<(), Error> {
+        let cancellation_token = self.state.cancellation_token.clone();
```

Comment:
The `clone` of a `clone` pattern here is a little confusing -- can you just clone `self.state.cancellation_token` directly when creating each task? That way you can also avoid having to disambiguate the local variable by adding `_clone` on the end.

Reply:
ah yeah that works